### PR TITLE
[FEAT] Entropic gw/fgw/srgw/srfgw solvers

### PR DESCRIPTION
**Pull request by cedricvincentcuaz**
## Types of changes

**changes to file ot/gromov/_bregman.py**:
- new features in `entropic_gromov_wasserstein`(2): 
i) `solver`: Set by default to `solver='PGD'` to not change the Projected Gradient Descent solver already implemented in the API. Add `solver='PPA'` to use the Proximal Point Algorithm from [50].
ii) `warmstart`: Either to perform warmstart of dual potentials in the successive Sinkhorn projections.
iii) `kwargs`: to pass parameters to `ot.sinkhorn` solver e.g to control its maximum number of iterations.

- new features in `entropic_gromov_barycenters`:
i) `warmstartT`: Either to perform warmstart of transport plans in the successive gromov-wasserstein transport problems.
ii) `kwargs`: to pass parameters to `ot.entropic_gromov_wasserstein` e.g using `solver='PGD'` or `solver='PPA'`.

- new functions: `entropic_fused_gromov_wasserstein`, `entropic_fused_gromov_wasserstein2`, `entropic_fused_gromov_barycenters` that extend existing functions for entropic GW to graphs with node features via FGW.

**changes to file ot/gromov/_gw.py**:
- new feature in `gromov_barycenters` and `fgw_barycenters`: 
i) `warmstartT`:  Either to perform warmstart of transport plans in the successive (fused) gromov-wasserstein transport problems.
ii) Fix bug in the update over structures when using `loss_fun='kl_loss'` 
- remove redundant function `update_structure_matrix` which was equivalent to `ot.gromov._utils.update_square_loss`

**changes to file ot/gromov/_semirelaxed.py**:
- new functions `entropic_semirelaxed_gromov_wasserstein`, `entropic_semirelaxed_gromov_wasserstein2`, `entropic_semirelaxed_fused_gromov_wasserstein`, `entropic_semirelaxed_fused_gromov_wasserstein2`: mirror descent algorithms following the KL geometry to solve semi-relaxed (Fused) Gromov-Wasserstein problems.

**changes to file ot/gromov/_utils.py**:
- move `ot.gromov._gw.update_feature_matrix` to `ot.gromov._utils.update_feature_matrix` as now use in both `_gw.py` and `_bregman.py`

**Nb : The marginals have been made optional for all solvers in these files.**
 
## Motivation and context / Related issue

Keep up with the (F)GW literature

## How has this been tested (if it applies)

**changes to file ot/gromov/_bregman.py**:
new tests: `test_entropic_proximal_gromov`, `test_entropic_proximal_fgw`, `test_entropic_fgw`, `test_asymmetric_entropic_fgw`, `test_entropic_fgw_dtype_device`, `test_entropic_fgw_barycenter`
modified tests: `test_entropic_gromov_dtype_device`, `test_entropic_fgw_dtype_device`, `test_gromov_entropic_barycenter`

**changes to file ot/gromov/_gw.py**:
modified tests: `test_gromov_barycenter`, `test_fgw_barycenter`

**changes to file ot/gromov/_semirelaxed.py**:
new tests: `test_entropic_semirelaxed_gromov`, `test_entropic_semirelaxed_gromov_dtype_device`, `test_entropic_semirelaxed_fgw`, `test_entropic_semirelaxed_fgw_dtype_device`

+ new test `test_not_implemented_solver` for entropic solvers and `loss_fun=kl_loss` for semi-relaxed solvers.
## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/gromov/_semirelaxed.py [hunk 1]
```diff
@@ -544,3 +544,487 @@ def solve_semirelaxed_gromov_linesearch(G, deltaG, cost_G, C1, C2, ones_p,
     cost_G = cost_G + a * (alpha ** 2) + b * alpha
 
     return alpha, 1, cost_G
+
+
+def entropic_semirelaxed_gromov_wasserstein(
+        C1, C2, p, loss_fun='square_loss', epsilon=0.1, symmetric=None,
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:550]: init p=None (uniform)

File: ot/gromov/_semirelaxed.py [hunk 2]
```diff
@@ -544,3 +544,487 @@ def solve_semirelaxed_gromov_linesearch(G, deltaG, cost_G, C1, C2, ones_p,
...
+        qG = nx.sum(G, 0)
+        marginal_product = nx.outer(ones_p, nx.dot(qG, fC2t))
+        log['srgw_dist'] = gwloss(constC + marginal_product, hC1, hC2, G, nx)
+        return G, log
+    else:
+        return G
+
+
+def entropic_semirelaxed_gromov_wasserstein2(
+        C1, C2, p, loss_fun='square_loss', epsilon=0.1, symmetric=None,
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:699]: same here

File: ot/gromov/_semirelaxed.py [hunk 3]
```diff
@@ -544,3 +544,487 @@ def solve_semirelaxed_gromov_linesearch(G, deltaG, cost_G, C1, C2, ones_p,
...
+        qG = nx.sum(G, 0)
+        marginal_product = nx.outer(ones_p, nx.dot(qG, fC2t))
+        log['srfgw_dist'] = alpha * gwloss(constC + marginal_product, hC1, hC2, G, nx) + (1 - alpha) * nx.sum(M * G)
+        return G, log
+    else:
+        return G
+
+
+def entropic_semirelaxed_fused_gromov_wasserstein2(
+        M, C1, C2, p, loss_fun='square_loss', symmetric=None, epsilon=0.1,
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:943]: again

**Review by rflamary — COMMENTED** [review]
very nice PR, next time do a smaller one but keep on with the quality code ;).

Small comments about usability below

**Comment by cedricvincentcuaz** [comment]
Thanks for the review Rémi :)

One thing I just realised. In entropic gw and fgw, there is no test on marginals if an initial transport plan G0 is given (different than None), should we add them ?
It would force us to recompute marginals at every iterations in the functions entropic (f)gw barycenter if `warmstartT=True`.

**Pull request merged**